core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2MetadataAsyncClient.java

```diff
@@ -0,0 +1,307 @@
+    }
+
+    @Override
+    public CompletableFuture<MetadataResponse> get(String path) {
+        return get(path, RetryPolicyContext.builder().retriesAttempted(0).build());
+    }
+
+    private CompletableFuture<MetadataResponse> get(String path, RetryPolicyContext retryPolicyContext) {
+        SdkHttpFullRequest baseTokenRequest = requestMarshaller.createTokenRequest(tokenTtl);
+
+        CompletableFuture<MetadataResponse> result = createAsyncRequest(baseTokenRequest)
+            .thenCompose(token -> {
+                SdkHttpFullRequest baseMetadataRequest = requestMarshaller.createDataRequest(path, token, tokenTtl);
+                return createAsyncRequest(baseMetadataRequest);
+            }).thenApply(MetadataResponse::create);
+
+        return result.handle((response, error) -> {
+            if (response != null || !shouldRetry(retryPolicyContext, error)) {
+                return result;
+            }
+            int newAttempt = retryPolicyContext.retriesAttempted() + 1;
+            log.debug(() -> "Retrying request: Attempt " + newAttempt);
+            RetryPolicyContext newContext =
+                RetryPolicyContext.builder()
+                                  .request(baseTokenRequest)
+                                  .retriesAttempted(newAttempt)
+                                  .exception(SdkClientException.create(error.getMessage(), error))
+                                  .build();
+            return scheduledRetryAttempt(path, newContext);
+        }).thenCompose(Function.identity()); // only java 12 has .exceptionallyCompose()
+    }
+
+    private CompletableFuture<String> createAsyncRequest(SdkHttpFullRequest baseRequest) {
+        StringAsyncResponseHandler responseHandler = new StringAsyncResponseHandler();
+        SdkHttpContentPublisher requestContentPublisher = new SimpleHttpContentPublisher(baseRequest);
+        AsyncExecuteRequest metadataRequest = AsyncExecuteRequest.builder()
+                                                                 .request(baseRequest)
+                                                                 .requestContentPublisher(requestContentPublisher)
+                                                                 .responseHandler(responseHandler)
+                                                                 .build();
+        httpClient.execute(metadataRequest);
+        return responseHandler.getResult();
+    }
+
+    private CompletableFuture<MetadataResponse> scheduledRetryAttempt(String path, RetryPolicyContext retryPolicyContext) {
+        Duration retryDelay = retryPolicy.backoffStrategy().computeDelayBeforeNextRetry(retryPolicyContext);
+        Executor retryExecutor = retryAttempt -> asyncRetryScheduler.schedule(retryAttempt, retryDelay.toMillis(),
+                                                                              TimeUnit.MILLISECONDS);
+        return CompletableFuture.supplyAsync(() -> get(path, retryPolicyContext), retryExecutor)
```

Comment:
(optional) you could supply the function to `scheduleRetryAttempt` as well 
`return scheduledRetryAttempt(() -> get(path, newContext), newContext)` or even 
`return scheduledRetryAttempt(() -> get(path, newContext), calculateRetryDelay(newContext))`